File: attentive-android-sdk/src/main/java/com/attentive/androidsdk/events/PurchaseEvent.java
package com.attentive.androidsdk.events;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.attentive.androidsdk.ParameterValidation;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
import java.util.List;

@JsonDeserialize(builder = PurchaseEvent.Builder.class)
public final class PurchaseEvent extends Event {
    private final List<Item> items;
    private final Order order;
    private final Cart cart;

    private PurchaseEvent(Builder builder) {
        items = builder.items;
        order = builder.order;
        cart = builder.cart;
    }

    @JsonPOJOBuilder(withPrefix = "")
    public static final class Builder {
        private final List<Item> items;
        private final Order order;
        private Cart cart;

        @JsonCreator
        public Builder(@JsonProperty("items") List<Item> items, @JsonProperty("order") Order order) {
            ParameterValidation.verifyNotEmpty(items, "items");
            ParameterValidation.verifyNotNull(order, "order");

            this.items = items;
            this.order = order;
        }

        public Builder cart(Cart val) {
            cart = val;
            return this;
        }

        public PurchaseEvent build() {
            return new PurchaseEvent(this);
        }
    }

    @NonNull
    public List<Item> getItems() {
        return items;
    }

    @NonNull
    public Order getOrder() {
        return order;
    }

    @Nullable
    public Cart getCart() {
        return cart;
    }
}

File: attentive-android-sdk/src/main/java/com/attentive/androidsdk/events/PurchaseEvent.kt
package com.attentive.androidsdk.events

import kotlinx.serialization.Serializable
import timber.log.Timber

/**
 * A purchase event. Fires when the user completes an order.
 *
 * One [PurchaseEvent] with N items produces N `Purchase` requests plus one `OrderConfirmed`
 * request on the backend.
 *
 * @property items The items in the order. Must be non-empty for the event to produce any requests.
 * @property order The order identifier.
 * @property cart The cart state at checkout. Optional.
 */
@Serializable
data class PurchaseEvent(
    val items: List<Item>,
    val order: Order,
    val cart: Cart? = null,
) : Event() {
    @Serializable
    class Builder(
        private val items: List<Item>,
        private val order: Order,
    ) {
        private var cart: Cart? = null

        fun cart(`val`: Cart?): Builder {
            cart = `val`
            return this
        }

        fun build(): PurchaseEvent {
            Timber.d("PurchaseEvent: items: $items, order: $order, cart: $cart")
            return PurchaseEvent(items, order, cart)
        }
    }
}
